src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,701 @@
+        // Remove deleted slots from partially available set only
+        // Note: Do NOT increment `version`, only remove from set 3
+        self.partially_available.remove_ranges(ranges);
+    }
+
+    /// Checks if there is any overlap between the given slot ranges and the fully available slots.
+    ///
+    /// This function checks if any of the provided slot ranges overlap with `fully_available` (set 2).
+    /// Returns true if there is at least one overlapping slot, false otherwise.
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges to check. Must be sorted and normalized.
+    ///
+    /// # Returns
+    ///
+    /// `true` if any overlap exists, `false` otherwise.
+    pub fn has_fully_available_overlap(&self, ranges: &[SlotRange]) -> bool {
+        self.fully_available.has_overlap(ranges)
+    }
+
+    /// Checks if all requested slots are available and returns version information.
+    ///
+    /// This function performs an optimized availability check:
+    /// - If sets 2 & 3 are empty and the input exactly matches set 1: returns current version (stable)
+    /// - Uses `union_relation` to check if ``local` ∪ `fully_available`` covers the query
+    /// - Returns special version values for partial or no coverage
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges to check. Must be sorted and normalized.
+    ///
+    /// # Returns
+    ///
+    /// - Current version (0..u32::MAX-2): Slots match exactly and are stable
+    /// - `SLOTS_TRACKER_UNAVAILABLE` (u32::MAX): Required slots are not available
+    /// - `SLOTS_TRACKER_UNSTABLE_VERSION` (u32::MAX-1): Slots available but partial/inexact match
+    pub fn check_availability(&self, ranges: &[SlotRange]) -> u32 {
+        // Fast path: If sets 2 & 3 are empty and input exactly matches set 1
+        if self.fully_available.is_empty()
+            && self.partially_available.is_empty()
+            && self.local == ranges
+        {
+            return self.version;
+        }
+
+        // Full check: Use union_relation to check coverage
+        match self.local.union_relation(&self.fully_available, ranges) {
+            CoverageRelation::Equals if self.partially_available.is_empty() => {
+                // Exact match and no partial slots
+                self.version
+            }
+            CoverageRelation::Equals | CoverageRelation::Covers => {
+                // Covered but not exact, or has partial slots
+                SLOTS_TRACKER_UNSTABLE_VERSION
+            }
+            CoverageRelation::NoMatch => {
+                // Not all slots are available
+                SLOTS_TRACKER_UNAVAILABLE
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn slot_ranges_slice(ranges: &[(u16, u16)]) -> &[SlotRange] {
+        unsafe { std::slice::from_raw_parts(ranges.as_ptr() as *const SlotRange, ranges.len()) }
+    }
```

Comment:
Rather implement the `From<(u16, u16)>` trait

Reply:
Using a `From<Range>` might also improve the DX

[Range](https://doc.rust-lang.org/std/ops/struct.Range.html)

Reply:
I implemented `From<(u16, u16)>`. For testing perpuses, I think seeing the numbers is the clearest way to understand the tests in this case